sec_certs/model/dependency_vulnerability_finder.py

```diff
@@ -0,0 +1,109 @@
+import logging
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import Dict, Optional, Set
+
+from sec_certs.sample.certificate import Certificate
+from sec_certs.serialization.json import ComplexSerializableType
+
+
+class DependencyType(Enum):
+    DIRECT = "direct"
+    INDIRECT = "indirect"
+
+
+@dataclass
+class DependencyCVE(ComplexSerializableType):
+    direct_dependency_cves: Optional[Set[str]] = field(default=None)
+    indirect_dependency_cves: Optional[Set[str]] = field(default=None)
+
+
+Certificates = Dict[str, Certificate]
+Vulnerabilities = Dict[str, Dict[str, Optional[Set[str]]]]
+
+
+class DependencyVulnerabilityFinder:
+    def __init__(self, certificates: Certificates):
+        self.vulnerabilities: Vulnerabilities = {}
+        self.certificates = certificates
+
+    def _get_dataset_cert_ids_occurrences(self) -> Dict[str, int]:
+        cert_id_occurrences: Dict[str, int] = {}
+
+        for dgst in self.certificates:
+            cert_id = self.certificates[dgst].heuristics.cert_id
+
+            if cert_id is None:
+                continue
+
+            cert_id_occurrences[cert_id] = cert_id_occurrences.get(cert_id, 0) + 1
+
+        return cert_id_occurrences
+
+    def _get_cert_dependency_cves(self, dgst: str, dependency_type: DependencyType) -> Optional[Set[str]]:
+        dependency_type_dict = {
+            DependencyType.DIRECT: self.certificates[dgst].heuristics.report_references.directly_referenced_by,
+            DependencyType.INDIRECT: self.certificates[dgst].heuristics.report_references.indirectly_referenced_by,
+        }
+
+        references = dependency_type_dict[dependency_type]
+
+        if not references:
+            return None
+
+        vulnerabilities = set()
+        dataset_cert_id_occurrences = self._get_dataset_cert_ids_occurrences()
+
+        for cert_id in references:
+            if cert_id is None:
+                continue
+
+            cert_id_occurrences = dataset_cert_id_occurrences.get(cert_id)
+
+            if cert_id_occurrences is None or cert_id_occurrences >= 2:
+                continue
+
+            for dgst in self.certificates:
+                cert_obj = self.certificates[dgst]
+                cert_obj_cves = cert_obj.heuristics.related_cves
+
+                if cert_obj.heuristics.cert_id == cert_id and cert_obj_cves:
+                    vulnerabilities.update(cert_obj_cves)
+
+        return vulnerabilities if vulnerabilities else None
+
+    def fit(self) -> Vulnerabilities:
```

Comment:
*Suggestion**: The `__init__()` should not take certificates as its parameter. Instead, they should be supplied via `fit` method. 

If you take a look at how sklearn classifiers work, they are instantiated only with parameters. The data is filled into them in `fit()` method, that should also owerwrite the previous state. 

By doing so, you can actually reuse the `DependencyVulnerabilityFinder` with different lists of certificates.